Sifive rvv intrinsic xsfvcp

yulong18 opened the pull request:
This pr tries add the intrinsics support for Xsfvcp extension.

yulong18 commented:
Hi, @kito-cheng :
Please help to review the code, thanks.
^_^

Liaoshihua committed:
RISC-V: Update Xsfvqmacc and Xsfvfnrclip's testcases

kito-cheng commented on gcc/config/riscv/constraints.md:
```diff
@@ -295,3 +295,13 @@
   (and (match_code "const_int")
        (match_test "IN_RANGE (ival, -64, -1)")))
+
+(define_constraint "B"
```
```suggestion
(define_constraint "Ou02"
```

kito-cheng commented on gcc/config/riscv/constraints.md:
```diff
@@ -295,3 +295,13 @@
+  (and (match_code "const_int")
+       (match_test "IN_RANGE (ival, 0, 3)")))
+
+(define_constraint "O"
```
```suggestion
(define_constraint "Ou01"
```

kito-cheng commented on gcc/config/riscv/constraints.md:
```diff
@@ -295,3 +295,13 @@
        (match_test "IN_RANGE (ival, -64, -1)")))
+
+(define_constraint "B"
+  "A 2-bit unsigned immediate for CSR access instructions."
```
```suggestion
  "A 2-bit unsigned immediate."
```

kito-cheng commented on gcc/config/riscv/constraints.md:
```diff
@@ -295,3 +295,13 @@
+       (match_test "IN_RANGE (ival, 0, 3)")))
+
+(define_constraint "O"
+  "A 1-bit unsigned immediate for CSR access instructions."
```
```suggestion
  "A 1-bit unsigned immediate."
```

kito-cheng reviewed (commented):
- I would like to use O[us][0-9][0-9] as a generic immediate operand modifier, so that it can used in other places as well, do you have intestest to post a proposal to riscv-c-api-doc like https://github.com/llvm/llvm-project/pull/112561

- could you add vsetvli check in the test case?
- We don't need X2 here, could you check how wadd.vv and wadd.vx implement? use double_trunc_vector should be work in this case once you adjust the base SEW/LMUL right.

kito-cheng commented on gcc/config/riscv/sifive-vector.md:
```diff
@@ -182,3 +182,599 @@
   [(set_attr "type" "sf_vfnrclip")
    (set_attr "mode" "<MODE>")])
+
+;; SF_VCP
+(define_insn "@sf_vc_x<mode>"
+  [(set (match_operand:VFULLI 0 "register_operand" "=vd")
+	(if_then_else:VFULLI
+	  (unspec:<VM>
+	    [(match_operand:<VM> 1 "vector_mask_operand"             "  Wc1")
+	     (match_operand 7 "vector_length_operand"                "   rK")
+	     (match_operand 8 "const_int_operand"                    "    i")
+	     (match_operand 9 "const_int_operand"                    "    i")
+	     (match_operand 10 "const_int_operand"                   "    i")
+	     (reg:SI VL_REGNUM)
+	     (reg:SI VTYPE_REGNUM)] UNSPEC_VPREDICATE)
+	  (unspec:VFULLI
+	    [(match_operand:SI 3 "const_int_operand" "B")
+	     (match_operand:SI 4 "const_int_operand" "K")
+	     (match_operand:SI 5 "const_int_operand" "K")
+	     (match_operand:<VEL> 6 "register_operand" "r")] UNSPEC_SF_CV)
+       (match_operand:VFULLI 2 "vector_merge_operand"     "vu")))]
+  "TARGET_VECTOR && TARGET_XSFVCP"
+  "sf.vc.x\t %3, %4, %5, %6"
+  [(set_attr "type" "sf_vc_se")
+   (set_attr "mode" "<MODE>")])
```
For those instruction without output, you don't really need a dummy output operand there, you just need some thing like

kito-cheng commented on gcc/config/riscv/sifive-vector.md:
```diff
@@ -182,3 +182,599 @@
+  [(set (match_operand:VFULLI 0 "register_operand" "=&vd")
+	(if_then_else:VFULLI
+	  (unspec:<VM>
+	    [(match_operand:<VM> 1 "vector_mask_operand"             "  Wc1")
+	     (match_operand 6 "vector_length_operand"                "   rK")
+	     (match_operand 7 "const_int_operand"                    "    i")
+	     (match_operand 8 "const_int_operand"                    "    i")
+	     (match_operand 9 "const_int_operand"                   "    i")
```
Same issue in following code
```suggestion
	    [(match_operand:<VM> 1 "vector_mask_operand"             "  Wc1")
	     (match_operand 6 "vector_length_operand"                "   rK")
	     (match_operand 7 "const_int_operand"                    "    i")
	     (match_operand 8 "const_int_operand"                    "    i")
	     (match_operand 9 "const_int_operand"                    "    i")
```

kito-cheng reviewed (commented):
Could you reduce the test files? just one test for each instruction is fine, you don't need put all test into gcc source tree.

e.g. only pick test_sf_vc_v_xvw_u8mf4 and test_sf_vc_v_xvw_se_u16m4 for sf.vc.v.xvw

pz9115 committed:
RISC-V: Update Xsfvfnrclip implementation.

gcc/ChangeLog:

	* config/riscv/genrvv-type-indexer.cc (expand_floattype): New func.
	(main):
	* config/riscv/riscv-vector-builtins-types.def (DEF_RVV_XFQF_OPS):
	(vint8mf8_t):
	(vint8mf4_t):
	(vint8mf2_t):
	(vint8m1_t):
	(vint8m2_t):
	* config/riscv/riscv-vector-builtins.cc (DEF_RVV_XFQF_OPS):
	(rvv_arg_type_info::get_xfqf_float_type):
	* config/riscv/riscv-vector-builtins.def (xfqf_vector):
	(xfqf_float):
	* config/riscv/riscv-vector-builtins.h (struct rvv_arg_type_info):
	* config/riscv/sifive-vector.md:
	* config/riscv/vector-iterators.md:

Liaoshihua committed:
RISC-V: Add intrinsic testcases for SiFive Xsfvcp extensions.

yulong18 committed:
RISC-V: Add intrinsics support for SiFive Xsfvcp extensions.
